lib/api/api_csrf.go

```diff
@@ -74,7 +74,7 @@ func (m *csrfManager) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	if strings.HasPrefix(r.URL.Path, "/rest/noauth") {
+	if strings.HasPrefix(r.URL.Path, "/rest/noauth") || r.URL.Path == "/rest/svc/lang" { // FIXME: this duplicates some logic in basicAuthAndSessionMiddleware in api_auth.go
```

Comment:
I tried to extract a function `isNoAuthPath(string)` from [`api_auth.go` line 103](https://github.com/syncthing/syncthing/pull/8757/files#diff-0d991cd6b91039a544e1e49f7a872291a13792e99fad033434f442e732d10607R103) and use that here, but that made the CSRF tests fail for some reason, and I can't figure out why. It does not seem to be a timing issue or race condition between parallel tests. Anyone have any ideas?

```
--- FAIL: TestCSRFRequired (0.02s)
    --- FAIL: TestCSRFRequired//rest_with_a_token_should_succeed (0.00s)
        api_test.go:947: Getting /rest/system/config with CSRF token should succeed, not 403 Forbidden
```

See commit https://github.com/emlun/syncthing/commit/f2ed03034f13bd68e28ee0b3da17d40c0b25e6a3 for a reproduction diff.

Reply:
I clearly don't understand the exact purpose/machanics of our CSRF measures, but the problem is likely best explained by this comment in api.go:

```
	// Wrap everything in CSRF protection. The /rest prefix should be
	// protected, other requests will grant cookies.
```

Basically the new/reused method catches and aborts the CSRF method for too many paths.

Reply:
Ah, it could be that aborting here prevents setting the CSRF cookie in the next section. I'll look into that. But I don't understand why the cookie should be set by other unprotected paths but not by `/rest/noauth` in particular...

Reply:
Yeah I don't get that either - I just looked into the test failure which this explains. And then briefly tried to understand the CSRF stuff without success (and didn't keep digging).

Reply:
I think I understand now. The test fails because it fails to initialize the CSRF cookie value by calling `GET /`, which is now covered by `isNoAuthPath()`, but the cookie is set only after that check if the URL path also does not start with the CSRF-protected prefix. The no-auth exception here was added in commit 7a402409f15f888e38ca0fef489d5ad369e8a9ac, in PR #8585.

The purpose of adding this exception was clearly to relax the CSRF policy. In that sense the order of the checks doesn't matter - there is no API path that would not have received a cookie before 7a402409f15f888e38ca0fef489d5ad369e8a9ac but not after 7a402409f15f888e38ca0fef489d5ad369e8a9ac.

So it seems safe to simply move the no-auth check to after the CSRF prefix check, if we compare behaviours before and after this PR:

- Before this PR:
  - Anything under `/rest/noauth` is allowed through without CSRF but will not receive a CSRF cookie.
  - Anything outside `/rest` receives a CSRF cookie.
- After this PR:
  - Anything outside `/rest` receives a CSRF cookie.
  - Anything outside `/rest` but on the no-auth exception list is allowed through without CSRF but will not receive a CSRF cookie. At this time, this is the exact path `/rest/svc/lang` and the path prefix `/rest/noauth`.

So the only effective difference is that `/rest/svc/lang` is now also allowed through without CSRF. This is needed for the language setting to work on the login page, and is safe as long as that is a readonly API path (i.e., with no side effects), which it currently is.

I've pushed these changes and removed the FIXME now. :slightly_smiling_face: If you want you can verify that the test begin to fail on 7c038fd8d68f943334fa944d199d31e16d195ca6, then emit a better failure message on c792ae8dad39ee6ddb81085b7eebeefef7a23f21, then is fixed by commit 797cd0e534417dbb56dd8c71b62cfd82a34b8d5f.